### PR TITLE
Fix OAuthHandler to use OriginalPathBase

### PR DESCRIPTION
**Pull request by HaoK**
Fixes Map path issue, and consolidates the logic into common helpers now

(Unfortunately can't write a test for twitter like fb since it doesn't surface the redirectUrl easily)

cc @Tratcher

### CODE REVIEW

File: src/Microsoft.AspNet.Authentication.OAuth/OAuthAuthenticationHandler.cs [hunk 1]
```diff
@@ -117,7 +117,7 @@ public OAuthAuthenticationHandler(HttpClient backchannel)
                 }
 
                 var requestPrefix = Request.Scheme + "://" + Request.Host;
-                var redirectUri = requestPrefix + RequestPathBase + Options.CallbackPath;
+                var redirectUri = requestPrefix + OriginalPathBase + Options.CallbackPath;
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Authentication.OAuth/OAuthAuthenticationHandler.cs]: Can you use your new BuildRedirectUri helper here?

**Comment by Tratcher** [comment]
:shipit:

File: test/Microsoft.AspNet.Authentication.Test/Facebook/FacebookMiddlewareTests.cs [hunk 2]
```diff
@@ -62,6 +64,37 @@ public class FacebookMiddlewareTests
         }
 
         [Fact]
+        public async Task MapWillNotAffectRedirect()
```
Comment (PinpointTownes) [inline test/Microsoft.AspNet.Authentication.Test/Facebook/FacebookMiddlewareTests.cs]: Is it worth adding another test where `app.UseFacebookAuthentication` and `app.Map` are both in a global `app.Map`?.

```csharp
var server = CreateServer(app => {
    app.Map("/base", map => {
        map.UseFacebookAuthentication();
        map.Map("/login", signoutApp => signoutApp.Run(context => context.Authentication.ChallengeAsync("Facebook", new AuthenticationProperties() { RedirectUri = "/" })));
     });
});
```

In this case, the `redirect_uri` should be `http://example.com/base/signin-facebook`.

File: src/Microsoft.AspNet.Authentication.OAuth/OAuthAuthenticationHandler.cs (hunk 1 above)
Comment (HaoK) [inline src/Microsoft.AspNet.Authentication.OAuth/OAuthAuthenticationHandler.cs]: yep will fix

File: test/Microsoft.AspNet.Authentication.Test/Facebook/FacebookMiddlewareTests.cs (hunk 2 above)
Comment (HaoK) [inline test/Microsoft.AspNet.Authentication.Test/Facebook/FacebookMiddlewareTests.cs]: Sure fair enough

**Comment by HaoK** [comment]
78cf7f99ffd0b18f5beac89ef37345e4aef66dc1

**Pull request closed without merging**